haystack/finder.py

```diff
@@ -96,3 +98,164 @@ def get_answers_via_similar_questions(self, question: str, top_k_retriever: int
             results["answers"].append(cur_answer)
 
         return results
+
+    def eval(self, label_index: str = "feedback", doc_index: str = "eval_document", label_origin: str = "gold_label",
+             top_k_retriever: int = 10, top_k_reader: int = 10):
+        """
+        Evaluation of the whole pipeline by first evaluating the Retriever and then evaluating the Reader on the result
+        of the Retriever.
+
+        Returns a dict containing the following metrics:
+            - "retriever_recall": Proportion of questions for which correct document is among retrieved documents
+            - "retriever_map": Mean of average precision for each question. Rewards retrievers that give relevant
+              documents a higher rank.
+            - "reader_recall": Proportion of predicted answers that overlap with correct answer
+            - "reader_map": Reader re-ranks documents retrieved by Retriever. Mean average precision of the re-ranking.
```

Comment:
Really nice, details docstrings! Just about the "reader_map" one: This might be confusing as the reader is not really re-ranking documents, right? It's more ranking the answers and they relate to documents. So maybe something like "Mean average precision of ranked answers by the Reader. Rewards readers that give correct answer(s) a high rank (mAP= 1 means all valid answers always on very top of list).

Reply:
Bogdan and me discussed this - a simple intuition for the user is to say the reader "re-ranks" retriever results, which is of course not the whole story for extractive QA. Also see line 186f where we actual count re ranking of pure docs in the reader_map. So lets either change this reader_map or increase complexity in the documentation like:

Reader re-ranks documents returned by Retriever. Mean average precision of the re-ranking. In case of extractive Readers the actual answer is compared to the extracted prediction. So in this case the Reader is not just re-ranking documents but comparing label and prediction on a different scale than the retriever.